cmd/relay/bgs/bgs.go

```diff
@@ -0,0 +1,1375 @@
+	}
+
+	if ban {
+		return nil, fmt.Errorf("cannot create user on pds with banned domain")
+	}
+
+	if peering.ID == 0 {
+		// why didn't we know about this PDS from requestCrawl?
+		// _maybe this never happens_ ? because of above peering Find ?
+		bgs.log.Warn("pds discovered in new user flow", "pds", durl.String(), "did", did)
+		pclient := &xrpc.Client{Host: durl.String()}
+		bgs.config.ApplyPDSClientSettings(pclient)
+		// TODO: the case of handling a new user on a new PDS probably requires more thought
+		cfg, err := comatproto.ServerDescribeServer(ctx, pclient)
+		if err != nil {
+			// TODO: failing this shouldn't halt our indexing
+			return nil, fmt.Errorf("failed to check unrecognized pds: %w", err)
+		}
+
+		// since handles can be anything, checking against this list doesn't matter...
+		_ = cfg
+
+		// TODO: could check other things, a valid response is good enough for now
+		peering.Host = durl.Host
+		peering.SSL = (durl.Scheme == "https")
+		peering.RateLimit = float64(bgs.slurper.DefaultPerSecondLimit)
+		peering.HourlyEventLimit = bgs.slurper.DefaultPerHourLimit
+		peering.DailyEventLimit = bgs.slurper.DefaultPerDayLimit
+		peering.RepoLimit = bgs.slurper.DefaultRepoLimit
+
+		if bgs.ssl && !peering.SSL {
+			return nil, fmt.Errorf("did references non-ssl PDS, this is disallowed in prod: %q %q", did, pdsService.URL)
+		}
+
+		if err := bgs.db.Create(&peering).Error; err != nil {
+			return nil, err
+		}
+	}
+
+	if peering.ID == 0 {
+		panic("somehow failed to create a pds entry?")
+	}
+
+	if peering.Blocked {
+		return nil, fmt.Errorf("refusing to create user with blocked PDS")
+	}
+
+	if peering.RepoCount >= peering.RepoLimit {
+		return nil, fmt.Errorf("refusing to create user on PDS at max repo limit for pds %q", peering.Host)
+	}
+
+	bgs.extUserLk.Lock()
+	defer bgs.extUserLk.Unlock()
+
+	user, err := bgs.lookupUserByDid(ctx, did)
+	if err == nil {
+		bgs.log.Debug("lost the race to create a new user", "did", did, "existing_hand", user.Handle)
+		if user.PDS != peering.ID {
+			// User is now on a different PDS, update
+			err = bgs.db.Transaction(func(tx *gorm.DB) error {
+				res := tx.Model(User{}).Where("id = ?", user.ID).Update("pds", peering.ID)
+				if res.Error != nil {
+					return fmt.Errorf("failed to update users pds: %w", res.Error)
+				}
+				res = tx.Model(&models.PDS{}).Where("id = ? AND repo_count < repo_limit", peering.ID).Update("repo_count", gorm.Expr("repo_count + 1"))
+				return nil
+			})
+
+			user.PDS = peering.ID
+		}
+		return user, nil
+	}
+
+	if !errors.Is(err, gorm.ErrRecordNotFound) {
+		return nil, err
+	}
+
+	// TODO: request this users info from their server to fill out our data...
+	u := User{
+		Did:         did,
+		PDS:         peering.ID,
+		ValidHandle: false,
+	}
+
+	err = bgs.db.Transaction(func(tx *gorm.DB) error {
+		res := tx.Model(&models.PDS{}).Where("id = ? AND repo_count < repo_limit", peering.ID).Update("repo_count", gorm.Expr("repo_count + 1"))
```

Comment:
I don't really understand the `repo_count < repo_limit` clause here... is this how account limits are enforced? the increment fails if the clause filters out the relevant row?

my intuition is that the account should be created (inserted), but with local account state "throttled", if the overall PDS is over quota. instead of silently being dropped; the later is hard for external folks to debug.

Reply:
I think this code got moved around, but this "throttled" state didn't get implemented yet.